### ISSUE TITLE
Cannot navigate back to a map if have launched a url from another page

### ISSUE THREAD

**Issue by admarwick** [flutter/flutter#26384]
Some kind of strange interaction between `url_launcher` and `google_maps_flutter` that means one cannot navigate back to a page with a map after launching a url from another page. But the problem only occurs on IOS; Android is OK.

Steps to reproduce:
1. See code below. Test app consists of two pages: (i) MapsDemo with a map (copied from the readme of google_map_flutter), and (ii) A page with link which launches a url via url_launcher.
2. Run the app. Do not press the Floating Action Button yet. 
3. Instead, press the raised button to move to the second page. On that page, click the link to launch a web page.
4. Verify that you see the web page, then press Done on the web browser, then press the back arrow to return to the first page. All OK so far.
5. Press the Floating Action Button on the first page to show the map. Then repeat step 3.
6. Verify that you cannot return to the map page by pressing the back button on the second page.
7. Lastly, verify that, after restarting the app and with the map showing, you can navigate to the second page and back without a problem, as long as you don't launch a url from the second page.

Luckily there's a work-around, which is to hide the map in the onPress handler of the button that navigates to the second page, and to re-show the map when the page route completes to return to the map page. However it would be nice if the IOS version worked as cleanly as the Android one does.

```dart
import 'package:flutter/material.dart';
import 'package:google_maps_flutter/google_maps_flutter.dart';
import 'package:url_launcher/url_launcher.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  // This widget is the root of your application.
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MapsDemo(),
    );
  }
}

class MapsDemo extends StatefulWidget {
  @override
  State createState() => MapsDemoState();
}

class MapsDemoState extends State<MapsDemo> {
  GoogleMapController mapController;

  bool _showMap = false;

  @override
  Widget build(BuildContext context) {
    print('MapsDemoState: build');
    return Scaffold(
        appBar: AppBar(title: Text('Test App')),
        floatingActionButton: FloatingActionButton(
          onPressed: () {
            setState(() => _showMap = !_showMap);
          },
          child: _showMap ? Icon(Icons.location_off) : Icon(Icons.location_on),
        ),
        body: !_showMap
            ? Center(
                child: RaisedButton(
                child: const Text('Second page'),
                onPressed: () {
                  Navigator.push(
                      context,
                      new MaterialPageRoute(
                          builder: (context) => SecondPage()));
                },
              ))
            : Padding(
                padding: EdgeInsets.all(15.0),
                child: Column(
                    mainAxisAlignment: MainAxisAlignment.spaceEvenly,
                    children: <Widget>[
                      Center(
                        child: SizedBox(
                          width: 300.0,
                          height: 200.0,
                          child: GoogleMap(
                            onMapCreated: _onMapCreated,
                          ),
                        ),
                      ),
                      RaisedButton(
                        child: const Text('Go to London'),
                        onPressed: mapController == null
                            ? null
                            : () {
                                mapController.animateCamera(
                                    CameraUpdate.newCameraPosition(
                                  const CameraPosition(
                                    bearing: 270.0,
                                    target: LatLng(51.5160895, -0.1294527),
                                    tilt: 30.0,
                                    zoom: 17.0,
                                  ),
                                ));
                              },
                      ),
                      RaisedButton(
                        child: const Text('Second page'),
                        onPressed: () {
                          Navigator.push(
                            context,
                            new MaterialPageRoute(
                                builder: (context) => SecondPage()),
                          );
                        },
                      ),
                    ]),
              ));
  }

  void _onMapCreated(GoogleMapController controller) {
    setState(() {
      mapController = controller;
    });
  }
}

class SecondPage extends StatefulWidget {
  @override
  State<StatefulWidget> createState() {
    return SecondPageState();
  }
}

class SecondPageState extends State<SecondPage> {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Second Page'),
      ),
      body: Center(
        child: Column(
            mainAxisAlignment: MainAxisAlignment.spaceEvenly,
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              InkWell(
                child: Text('Launch web page'),
                onTap: () => launch('https://flutter.io'),
              ),
            ]),
      ),
    );
  }
}
```
Flutter doctor:
```
[✓] Flutter (Channel beta, v1.0.0, on Mac OS X 10.13.6 17G4015, locale en-US)
    • Flutter version 1.0.0 at /Users/alan/Documents/util/flutter
    • Framework revision 5391447fae (6 weeks ago), 2018-11-29 19:41:26 -0800
    • Engine revision 7375a0f414
    • Dart version 2.1.0 (build 2.1.0-dev.9.4 f9ebf21297)

[✓] Android toolchain - develop for Android devices (Android SDK 28.0.3)
    • Android SDK at /Users/alan/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.3
    • ANDROID_HOME = /Users/alan/Library/Android/sdk
    • Java binary at: /Applications/Android Studio 3.2.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 2.0.0
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.2)
    • Android Studio at /Applications/Android Studio 3.2.app/Contents
    • Flutter plugin version 31.3.1
    • Dart plugin version 181.5656
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-1136-b06)

[✓] Connected device (1 available)
    • Alan’s iPhone 5S • fb2c341f9cecedf031922c7a9f3977426a476bf6 • ios • iOS 12.1.2

• No issues found!

```

**Comment by admarwick** [comment]
It turns out that the "work around" I mentioned in the issue report is actually quite a pain, as when a map is removed from the widget tree it loses a lot of state, which has to be re-instated when the map is re-shown. More platform-specific code.

**Comment by ozzy91** [comment]
I ran into the same issue in my app. I also noticed it with url_launcher, but after further testing I found out that it's not directly related.

It seems to be an issue with the AppLifecycle in general. Instead of launching a url, just hit the home button and then go back to the app.
Closing a SafariViewController from url_launcher triggers the same `didChangeAppLifecycleState` callback with state `AppLifecycleState.resumed` as does returning to the app from a paused state.
